sklearn/metrics/pairwise.py

```diff
@@ -992,24 +992,28 @@ def detect_categorical_features(X, categorical_features=None):
 
     Returns
     -------
-    categorical_features_mask, numerical_features_mask
-     : ndarray, shape (n_features)
+    categorical_features_mask : ndarray, shape (n_features)
+
     """
-    _, n_cols = np.shape(X)
     # Automatic detection of categorical features
     if categorical_features is None:
-        categorical_features = np.zeros(n_cols, dtype=bool)
-        for col in range(0, n_cols):
-            if not np.issubdtype(type(X[0][col]), np.number):
+        categorical_features = np.zeros(np.shape(X)[1], dtype=bool)
+
+        def detect_cat(x):
+            return x != np.nan and not np.issubdtype(type(x), np.number)
+
+        for col in range(np.shape(X)[1]):
+            f_test = np.frompyfunc(detect_cat, 1, 1)
+            if np.any(f_test(X[:, col])):
```

Comment:
If this is a pandas DataFrame, we need to be using X.iloc here. Consider using our safe_indexing util

Please test with a DataFrame whose columns are named.

Reply:
There is no need, we just use np.ndarray at this stage.